pkg/volume/downwardapi/downwardapi_test.go

```diff
@@ -493,346 +228,131 @@ func TestWriteTwiceWithUpdate(t *testing.T) {
-		t.Errorf(err.Error())
+	if err := unmounter.TearDown(); err != nil {
+		test.t.Errorf("Expected success, got: %v", err)
 	}
-
-	if sortLines(string(data)) != sortLines(fieldpath.FormatMap(newLabels)) {
-		t.Errorf("Found `%s` expected %s", data, fieldpath.FormatMap(newLabels))
+	if _, err := os.Stat(test.volumePath); err == nil {
+		test.t.Errorf("TearDown() failed, volume path still exists: %s", test.volumePath)
+	} else if !os.IsNotExist(err) {
+		test.t.Errorf("SetUp() failed: %v", err)
 	}
-	CleanEverything(plugin, testVolumeName, volumePath, testPodUID, t)
+	os.RemoveAll(test.rootDir)
 }
 
-func TestWriteWithUnixPath(t *testing.T) {
-	var (
-		testPodUID     = types.UID("test_pod_uid")
-		testVolumeName = "test_write_with_unix_path"
-		testNamespace  = "test_metadata_namespace"
-		testName       = "test_metadata_name"
-	)
-
-	labels := map[string]string{
-		"key1": "value1",
-		"key2": "value2",
-		"key3": "value3\n"}
-
-	annotations := map[string]string{
-		"a1": "value1",
-		"a2": "value2"}
-
-	clientset := fake.NewSimpleClientset(&v1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      testName,
-			Namespace: testNamespace,
-			Labels:    labels,
-		},
-	})
-
-	pluginMgr := volume.VolumePluginMgr{}
-	tmpDir, host := newTestHost(t, clientset)
-	defer os.RemoveAll(tmpDir)
-	pluginMgr.InitPlugins(ProbeVolumePlugins(), host)
-	plugin, err := pluginMgr.FindPluginByName(downwardAPIPluginName)
-	defaultMode := int32(0644)
-	volumeSpec := &v1.Volume{
-		Name: testVolumeName,
-		VolumeSource: v1.VolumeSource{
-			DownwardAPI: &v1.DownwardAPIVolumeSource{
-				DefaultMode: &defaultMode,
-				Items: []v1.DownwardAPIVolumeFile{
-					{Path: "this/is/mine/labels", FieldRef: &v1.ObjectFieldSelector{
-						FieldPath: "metadata.labels"}},
-					{Path: "this/is/yours/annotations", FieldRef: &v1.ObjectFieldSelector{
-						FieldPath: "metadata.annotations"}},
-				}}},
-	}
-	if err != nil {
-		t.Errorf("Can't find the plugin by name")
-	}
-	pod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{UID: testPodUID, Labels: labels, Annotations: annotations}}
-	mounter, err := plugin.NewMounter(volume.NewSpecFromVolume(volumeSpec), pod, volume.VolumeOptions{})
-
-	if err != nil {
-		t.Errorf("Failed to make a new Mounter: %v", err)
-	}
-	if mounter == nil {
-		t.Errorf("Got a nil Mounter")
-	}
-
-	volumePath := mounter.GetPath()
-	err = mounter.SetUp(nil)
-	if err != nil {
-		t.Errorf("Failed to setup volume: %v", err)
-	}
-
-	var data []byte
-	data, err = ioutil.ReadFile(path.Join(volumePath, "this/is/mine/labels"))
-	if err != nil {
-		t.Errorf(err.Error())
-	}
+type testStep struct {
+	name             string
+	expectedMap      map[string]string
+	expectedText     string
+	expectedMode     int32
+	linkShouldChange bool
+	newLabels        map[string]string
+}
 
-	if sortLines(string(data)) != sortLines(fieldpath.FormatMap(labels)) {
-		t.Errorf("Found `%s` expected %s", data, fieldpath.FormatMap(labels))
+func (s testStep) run(test *downwardAPITest) {
+	switch {
```

Comment:
This block is fragile and is heavily dependent on the order.

Someone who is making changes to the existing tests if accidentally sets/unsets one of the variables above, then we can get unexpected behavior because of the ordering and dependence of a verification function on other unrelated factors.

I'd suggest working on a different approach for handling this. Either you can have a map with the testName as key and the verification function as value or something else on similar lines.